File: frontend/src/types/portfolio.ts
export interface PortfolioLot {
  id: string
  asset_id: string
  quantity: number
  unit_cost_usd: string
  purchased_on: string
  created_at: string
  updated_at: string
}

export interface PortfolioPosition {
  asset_id: string
  name: string
  set_name: string | null
  card_number: string | null
  game: string
  quantity: number
  average_unit_cost_usd: string
  cost_basis_usd: string
  latest_raw_price_usd: string | null
  latest_price_at: string | null
  market_value_usd: string | null
  unrealized_pnl_usd: string | null
  unrealized_pnl_percent: string | null
  valuation_status: 'priced' | 'unpriced'
  lots: PortfolioLot[]
}

export interface PortfolioAllocation {
  game: string
  market_value_usd: string
  percentage: string
  priced_position_count: number
}

export interface PortfolioSummary {
  total_cost_basis: string
  priced_cost_basis: string
  total_market_value: string
  unrealized_pnl: string
  unrealized_pnl_percent: string | null
  position_count: number
  priced_position_count: number
  unpriced_position_count: number
  valuation_coverage_percent: string
  position_limit: number | null
}

export interface Portfolio {
  summary: PortfolioSummary
  allocations: PortfolioAllocation[]
  positions: PortfolioPosition[]
}

export interface PortfolioLotInput {
  asset_id: string
  quantity: number
  unit_cost_usd: string
  purchased_on: string
}

export type PortfolioLotPatch = Partial<
  Pick<PortfolioLotInput, 'quantity' | 'unit_cost_usd' | 'purchased_on'>
>

export interface PortfolioApiErrorDetail {
  code?: string
  message?: string
  position_limit?: number
  position_count?: number
  upgrade_url?: string
}

File: backend/app/schemas/portfolio.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PortfolioLotCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    asset_id: UUID
    quantity: int = Field(ge=1)
    unit_cost_usd: Decimal = Field(ge=0, max_digits=12, decimal_places=2)
    purchased_on: date


class PortfolioLotPatchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    quantity: int | None = Field(default=None, ge=1)
    unit_cost_usd: Decimal | None = Field(
        default=None,
        ge=0,
        max_digits=12,
        decimal_places=2,
    )
    purchased_on: date | None = None

    @model_validator(mode="after")
    def require_supplied_non_null_field(self) -> PortfolioLotPatchRequest:
        if not self.model_fields_set:
            raise ValueError("at least one field must be supplied")
        if any(getattr(self, field) is None for field in self.model_fields_set):
            raise ValueError("supplied fields cannot be null")
        return self


class PortfolioLotResponse(BaseModel):
    id: UUID
    asset_id: UUID
    quantity: int
    unit_cost_usd: Decimal
    purchased_on: date
    created_at: datetime
    updated_at: datetime


class PortfolioPositionResponse(BaseModel):
    asset_id: UUID
    name: str
    set_name: str | None
    card_number: str | None
    game: str
    quantity: int
    average_unit_cost_usd: Decimal
    cost_basis_usd: Decimal
    latest_raw_price_usd: Decimal | None
    latest_price_at: datetime | None
    market_value_usd: Decimal | None
    unrealized_pnl_usd: Decimal | None
    unrealized_pnl_percent: Decimal | None
    valuation_status: Literal["priced", "unpriced"]
    lots: list[PortfolioLotResponse]


class PortfolioAllocationResponse(BaseModel):
    game: str
    market_value_usd: Decimal
    percentage: Decimal
    priced_position_count: int


class PortfolioSummaryResponse(BaseModel):
    total_cost_basis: Decimal
    priced_cost_basis: Decimal
    total_market_value: Decimal
    unrealized_pnl: Decimal
    unrealized_pnl_percent: Decimal | None
    position_count: int
    priced_position_count: int
    unpriced_position_count: int
    valuation_coverage_percent: Decimal
    position_limit: int | None


class PortfolioResponse(BaseModel):
    summary: PortfolioSummaryResponse
    allocations: list[PortfolioAllocationResponse]
    positions: list[PortfolioPositionResponse]
